### PR TITLE
Add IGListCollectionContext API to get visible indexes.

### PR DESCRIPTION
**Pull request by Malecks**
## Changes in this pull request
Add IGListCollectionContext API to get visible indexes:
`- (NSArray<NSIndexPath *> *)visiblePathsForSectionController:(IGListSectionController<IGListSectionType> *) sectionController`

Issue fixed: #465 

### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: Source/IGListAdapter.m [hunk 1]
```diff
@@ -857,6 +857,19 @@ - (__kindof UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index
     return cells;
 }
 
+- (NSArray<NSIndexPath *> *)visiblePathsForSectionController:(IGListSectionController<IGListSectionType> *) sectionController {
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: nit: can we name this visible**index**PathsForSectionController

File: Source/IGListAdapter.m [hunk 2]
```diff
@@ -857,6 +857,19 @@ - (__kindof UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index
...
+    NSMutableArray *paths = [NSMutableArray new];
+    UICollectionView *collectionView = self.collectionView;
+    NSArray *visiblePaths = [collectionView indexPathsForVisibleItems];
+    const NSInteger section = [self sectionForSectionController:sectionController];
+    for (NSIndexPath *path in visiblePaths) {
+        if (path.section == section) {
+            [paths addObject:path];
+        }
+    }
+    return paths;
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: Looks great!

File: Source/IGListCollectionContext.h [hunk 3]
```diff
@@ -59,6 +59,16 @@ NS_ASSUME_NONNULL_BEGIN
 - (NSArray<UICollectionViewCell *> *)visibleCellsForSectionController:(IGListSectionController<IGListSectionType> *)sectionController;
 
 /**
+ Returns the visible paths for the given section controller.
+ 
+ @param sectionController The section controller requesting this information.
+
```
Comment (rnystrom) [inline Source/IGListCollectionContext.h]: Kill whitespace

File: Source/IGListCollectionContext.h [hunk 4]
```diff
@@ -59,6 +59,16 @@ NS_ASSUME_NONNULL_BEGIN
 - (NSArray<UICollectionViewCell *> *)visibleCellsForSectionController:(IGListSectionController<IGListSectionType> *)sectionController;
 
 /**
+ Returns the visible paths for the given section controller.
+ 
+ @param sectionController The section controller requesting this information.
+ 
+ @return An array of visible index paths, or an empty array if none are found.
+
```
Comment (rnystrom) [inline Source/IGListCollectionContext.h]: This whitespace too

File: Source/IGListStackedSectionController.m [hunk 5]
```diff
@@ -187,6 +187,18 @@ - (UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index sectionController:
     return cells;
 }
 
+- (NSArray<NSIndexPath *> *)visiblePathsForSectionController:(IGListSectionController<IGListSectionType> *)sectionController {
+    NSMutableArray *paths = [NSMutableArray new];
+    id<IGListCollectionContext> collectionContext = self.collectionContext;
+    NSArray *visiblePaths = [collectionContext visiblePathsForSectionController:self];
+    for (NSIndexPath *path in visiblePaths) {
+        if (self.sectionControllersForItems[path.section] == sectionController) {
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: Hmm I think this should actually be `path.item` instead. A unit test would help to confirm.

**Comment by Malecks** [comment]
@rnystrom thanks! will fix up and submit tomorrow. 👍

File: Tests/IGListAdapterTests.m [hunk 6]
```diff
@@ -313,6 +313,21 @@ - (void)test_whenCellsExtendBeyondBounds_thatVisibleCellsExistForSectionControll
...
+    id sectionController3 = [self.adapter sectionControllerForObject:@3];
+    id sectionController4 = [self.adapter sectionControllerForObject:@4];
+    id sectionController5 = [self.adapter sectionControllerForObject:@5];
+    id sectionController6 = [self.adapter sectionControllerForObject:@6];
+    XCTAssertEqual([self.adapter visibleIndexPathsForSectionController:sectionController2].count, 2);
+    XCTAssertEqual([self.adapter visibleIndexPathsForSectionController:sectionController3].count, 3);
+    XCTAssertEqual([self.adapter visibleIndexPathsForSectionController:sectionController4].count, 4);
+    XCTAssertEqual([self.adapter visibleIndexPathsForSectionController:sectionController5].count, 1);
+    XCTAssertEqual([self.adapter visibleIndexPathsForSectionController:sectionController6].count, 0);
+}
```
Comment (rnystrom) [inline Tests/IGListAdapterTests.m]: Mind adding a test for the stacked section controller too? Make sure we're totally covered.

File: Source/IGListStackedSectionController.m [hunk 7]
```diff
@@ -187,10 +187,10 @@ - (UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index sectionController:
...
 }
 
-- (NSArray<NSIndexPath *> *)visiblePathsForSectionController:(IGListSectionController<IGListSectionType> *)sectionController {
+- (NSArray<NSIndexPath *> *)visibleIndexPathsForSectionController:(IGListSectionController<IGListSectionType> *)sectionController {
     NSMutableArray *paths = [NSMutableArray new];
     id<IGListCollectionContext> collectionContext = self.collectionContext;
-    NSArray *visiblePaths = [collectionContext visiblePathsForSectionController:self];
+    NSArray *visiblePaths = [collectionContext visibleIndexPathsForSectionController:self];
     for (NSIndexPath *path in visiblePaths) {
         if (self.sectionControllersForItems[path.section] == sectionController) {
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: Note: I still think there's a bug here. Should be using `path.item`

File: Source/IGListAdapter.m [hunk 8]
```diff
@@ -857,7 +857,7 @@ - (__kindof UICollectionViewCell *)cellForItemAtIndex:(NSInteger)index
     return cells;
 }
 
-- (NSArray<NSIndexPath *> *)visiblePathsForSectionController:(IGListSectionController<IGListSectionType> *) sectionController {
+- (NSArray<NSIndexPath *> *)visibleIndexPathsForSectionController:(IGListSectionController<IGListSectionType> *) sectionController {
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: Make sure you update the `CHANGELOG.md` now that we renamed this method 😄

File: Tests/IGListAdapterTests.m (hunk 6 above)
Comment (Malecks) [inline Tests/IGListAdapterTests.m]: I'll be honest I've just been following `visibleCellsForSectionController:` so not totally sure how to add tests for the stacked SC, but I'm working on it!

File: Source/IGListStackedSectionController.m (hunk 7 above)
Comment (Malecks) [inline Source/IGListStackedSectionController.m]: I did experiment with `path.item` but the tests I had added were failing. It's possible that I don't understand what's going on well enough to know whether they should be failing or not though 🤔

File: Tests/IGListAdapterTests.m (hunk 6 above)
Comment (rnystrom) [inline Tests/IGListAdapterTests.m]: @Malecks np, apologies that some of the `starter-task`s can have a bit more to them under the surface 😄 ! 

Off the top of my head:

- Copy [this test](https://github.com/Instagram/IGListKit/blob/master/Tests/IGListStackSectionControllerTests.m#L399-L417)
- Assert the actual `NSIndexPath` values (not just counts) for each child section controller

Hopefully that helps!

File: Tests/IGListStackSectionControllerTests.m [hunk 9]
```diff
@@ -416,6 +416,26 @@ - (void)test_whenQueryingVisibleSectionControllers_withCellsOffscreen_thatOnlyVi
...
+    IGListTestSection *section3 = stack.sectionControllers[2];
+    IGListTestSection *section4 = stack.sectionControllers[3];
+    IGListTestSection *section5 = stack.sectionControllers[4];
+    
+    XCTAssertEqual([self.adapter visibleIndexPathsForSectionController:stack].count, 10);
+    XCTAssertEqual([stack visibleIndexPathsForSectionController:section1].count, 3);
+    XCTAssertEqual([stack visibleIndexPathsForSectionController:section2].count, 4);
+    XCTAssertEqual([stack visibleIndexPathsForSectionController:section3].count, 0);
+    XCTAssertEqual([stack visibleIndexPathsForSectionController:section4].count, 3);
+    XCTAssertEqual([stack visibleIndexPathsForSectionController:section5].count, 0);
```
Comment (rnystrom) [inline Tests/IGListStackSectionControllerTests.m]: Last request, promise!

Can we check the actual values here and not just the counts? That'll be a really accurate test! It'll be a bit long-winded, but a more thorough test will withstand changes better.

Here's what I'm thinking:

```objc
// use sets b/c we don't care about order
NSSet *visible1 = [NSSet setWithArray:[stack visibleIndexPathsForSectionController:stack.sectionControllers[0]]];
// repeat for all 5

NSSet *expected1 = [NSSet setWithArray:@[
  [NSIndexPath indexPathForItem:0 inSection:0],
  // etc
]];
// repeat for all 5

XCTAssertEqualObjects(visible1, expected1);
// repeat for all 5
```

Make sense?

Comment (Malecks) [inline Tests/IGListStackSectionControllerTests.m]: Definitely. Good idea! 👍

**Comment by jessesquires** [comment]
Thanks again @Malecks ! 👍

**Pull request closed without merging**